<a href="https://colab.research.google.com/github/IrfanHilman/FTDS_Hacktiv8_Batch_16/blob/main/h8p0lc2_batch_irfan.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem 1

In [13]:
#Import library numpy sebagai np dan pandas sebagai pf
import numpy as np
import pandas as pd

In [11]:
#Buat variabel harga Rumah 1 dan Rumah 2
harga_rumah1 = 800000000
harga_rumah2 = 600000000

#Buat vektor Rumah 1 dan Rumah 2 menggunakan numpy dan tipe data array
rumah1 = np.array([100,240,3,1,1,0,1,1])
rumah2 = np.array([120,200,2,1,1,1,1,1])

In [12]:
#Hitung nilai dot product dan cosine similarity dari Rumah 1 dan Rumah 2
dot_product = rumah1 @ rumah2
cosine_sim = dot_product/(np.linalg.norm(rumah1)*np.linalg.norm(rumah2))
print('Dot product:', dot_product)
print('Cosine similarity:', cosine_sim)

Dot product: 60010
Cosine similarity: 0.9894028379856452


Jawaban Soal:

a. Berdasarkan nilai cosine similarity-nya, Rumah 1 dan Rumah 2 memiliki 98.94% kemiripan. Sehingga rumah yang seharusnya dibeli adalah Rumah 2 karena harganya lebih murah tetapi memiliki kualitas yang mirip dengan Rumah 1.

b. Berdasarkan definisinya, cos-1(0)= 90 derajat dan cos-1(1)= 0 derajat. Jadi jika vektor A dan B memiliki nilai cosine similarity mendekati 0, maka vektor A dan B akan tegak lurus satu sama lain di koordinat kartesian. Sebaliknya, jika vektor A dan B memiliki nilai cosine similarity mendekati 1, maka vektor A dan B akan sejajar dan searah satu sama lain, dengan kata lain, vektor A dan B memiliki kemiripan lebih tinggi.

c. Cosine similarity melibatkan vektor, bukan matrix, karena konsepnya mengukur kemiripan arah vektor dalam 2 dimensi sedangakan matrix melebihi 2 dimensi.

# Problem 2



Code yang digunakan untuk melakukan query dari Google Cloud adalah:

SELECT price

FROM bigquery-public-data.properati_properties_ar.properties_rent_201501

WHERE operation = 'rent'

AND property_type = 'apartment'

In [55]:
#Call data properties yang telah di-upload ke github. Assign data ke variabel df
df = pd.read_csv('https://github.com/IrfanHilman/FTDS_Hacktiv8_Batch_16/blob/main/h8dsft_P0LC3_Muhammad%20Irfan_Hilman.csv?raw=true')

In [56]:
#Tampilkan 5 baris data teratas
df.head()

price
0   2500
1   3500
2   2850
3   1850
4   1000

In [57]:
#Tampilkan 5 baris data terbawah
df.tail()

price
1653    500
1654    500
1655    500
1656    500
1657    500

1. Perhitungan central tendency (mean, median, modus) sebelum adanya data anomaly

In [58]:
#Hitung mean dari data price
mean = df.price.mean()
print('Mean dari data price adalah:', mean)

Mean dari data price adalah: 5166.355247285886


In [59]:
#Hitung median dari data price
median = df.price.median()
print('Median dari data price adalah:', median)

Median dari data price adalah: 2700.0


In [60]:
#Hitung modus dari data price
modus = df.price.mode()
modus = np.array(modus)
print('Modus dari data price adalah: %s' % str(modus[0]))

Modus dari data price adalah: 0


2. Check skewness data untuk mengetahui apakah datanya terdistribusi normal atau tidak

In [61]:
#Hitung skewness dari data price
skewness = df.price.skew()
print('Skewness dari data price adalah:', skewness)

Skewness dari data price adalah: 4.2401362864090695


Karena skewness dari data adalah positif (4.24), maka distribusi data adalah positive skewness dengan nilai mean (5166) lebih besar daripada median (2700)

3. Extreme Value Analysis

Analisa menggunakan Tukey's rule karena distibusi data adalah positive skewness

In [62]:
#Hitung Q1 (25th quantile), Q2 (50th quantile) dan Q3(75th quantile) dari data price
price = df.price
quartiles = np.percentile(price, [25, 50, 75])
Q1 = quartiles[0]
Q2 = quartiles[1]
Q3 = quartiles[2]
print('25th quantile adalah:', Q1)
print('50th quantile adalah:', Q2)
print('75th quantile adalah:', Q3)

25th quantile adalah: 0.0
50th quantile adalah: 2700.0
75th quantile adalah: 5500.0


In [71]:
#Hitung IQR = Q3-Q1
IQR = Q3-Q1
IQR

5500.0

In [72]:
#Hitung nilai batas atas Tukey dalam variabel Tukey_upper
Tukey_upper = Q3 + (1.5*IQR)
print('Tukey upper:', Tukey_upper)

#Hitung nilai batas bawah Tukey dalam variabel Tukey_lower
Tukey_lower = Q1 - (1.5*IQR)
print('Tukey lower:', Tukey_lower)

Tukey upper: 13750.0
Tukey lower: -8250.0


4. Buat variabel baru 'df_no_outliers' yang berisi data tanpa outliers berdasarkan Tukey's rule

In [70]:
#Buat variabel 'df_no_outliers' berisi data tanpa outliers
df_no_outliers = df[(df['price']<13750) & (df['price']>-8250)]
df_no_outliers

price
0      2500
1      3500
2      2850
3      1850
4      1000
...     ...
1653    500
1654    500
1655    500
1656    500
1657    500

[1500 rows x 1 columns]

In [73]:
#Hitung skewness pada data tanpa outliers
skewness_no_outliers = df_no_outliers.price.skew()
print('Nilai skewness data tanpa outliers adalah:', skewness_no_outliers)

Nilai skewness data tanpa outliers adalah: 1.3384880750664294


Jawaban Soal:

1. Sebelum dihilangkan outliers-nya, nilai rata-rata price adalah 5166, nilai median price adalah 2700 dan nilai modus price adalah 0. Kecenderungan pemusatan datanya adalah positive skewness dengan nilai skewness 4.24. Positive skewness ditandai dengan nilai mean > nilai median.

2. Setelah melakukan Extreme Value Analysis (EVA) atau penghilangan data outliers, nilai skewness data berkurang menjadi 1.34484. Artinya kecenderungan pemusatan datanya mendekati bentuk distribusi normal, dibandingkan sebelumnya, karena pada distribusi normal, skewness mendekati 0.

3. Terdapat 2 teknik dalam EVA: Tukey's rule dan Z-score. Jika distribusi data tidak normal (skew), Tukey's rule dapat digunakan. Kriteria data outliers pada Tukey's rule adalah data berada diatas Q3 + 1.5*IQR dan dibawah Q1 - 1.5*IQR dengan IQR = Q3-Q1. Sebaliknya, aturan Z-score hanya digunakan jika distribusi data adalah normal. Data yang masuk kriteria outliers dalam aturan Z-score adalah berada diatas mean + 3*Zscore dan dibawah mean -3*Zscore. Teknik yang saya pakai dalam Problem 2 ini adalah Tukey's rule, karena bentuk persebaran data adalah positive skewness (skewness>>0).